tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,396 @@
+      Module& M, SET* keyNodes,
+      GRAPH* schemaStrToFunctions, GRAPH* functionsToSchemaStrs) {
+    for (GlobalVariable& GV : M.globals()) {
+      if (!GV.hasInitializer()) continue;
+      Constant* gvInit = GV.getInitializer();
+      if (gvInit->isNullValue()) continue;
+      ConstantDataArray* gvInitArray = dyn_cast<ConstantDataArray>(gvInit);
+      if (!gvInitArray || !gvInitArray->isCString()) continue;
+      std::string gvInitStr = gvInitArray->getAsCString().str();
+      if (!isFunctionSchemaString(gvInitStr)) continue;
+      std::string schemaStr = truncateFunctionSchemaString(gvInitStr);
+
+      // Track all seen schema strings. Useful for checking whether we miss any
+      // dependency in the logic below.
+      keyNodes->insert(schemaStr);
+
+      // Search for GV users.
+      for (auto gvUser : GV.users()) {
+        // Search for "i8* getelementptr inbounds" ConstantExpr.
+        ConstantExpr* CE = dyn_cast<ConstantExpr>(gvUser);
+        if (!CE) continue;
+
+        // Search for CE users.
+        for (auto ceUser : CE->users()) {
+          // Search for "std::basic_string" constructor.
+          Instruction* ceUserInst = dyn_cast<Instruction>(ceUser);
+          if (!ceUserInst) continue;
+          analyzeFunctionSchemaUser(
+              schemaStr, *ceUserInst, schemaStrToFunctions,
+              functionsToSchemaStrs);
+        }
+      }
+    }
+  }
+
+  // Use fuzzy match to find all llvm::Function instances related to the
+  // function schema string. There are two types of relationship:
+  // 1) function schema to function (op registration);
+  // 2) function to function schema (op invocation);
+  //
+  // For op invocation, simply check whether the parent function calls
+  // 'c10::Dispatcher::findSchema' since the use of the schema string.
+  //
+  // For op registration, collect all function references between two following
+  // "c10::RegisterOperators::Options::schema" calls. The function references
+  // can be direct function call, function pointer or template argument.
+  //
+  // Note that these rules highly depend on how LLVM emit IRs. If it changes
+  // the sequence of schema string reference and c10::RegisterOperators API call
+  // then we might need use more sophisticated alias analysis, analyze clang-AST
+  // instead or simplify leverage the connection between function schema string
+  // and related function name (as they are generated by script).
+  static void analyzeFunctionSchemaUser(
+      const std::string& schemaStr,
+      Instruction& I,
+      GRAPH* schemaStrToFunctions,
+      GRAPH* functionsToSchemaStrs) {
+    std::string parentFunctionName = I.getFunction()->getName();
+    bool seenOptionsSchemaCall = false;
+    for (Instruction* cur = getNextInstruction(I);
+         cur;
+         cur = getNextInstruction(*cur)) {
+      // Check pattern to call registered ops.
+      if (checkInstructionCalledFunction(*cur, "c10::Dispatcher::findSchema")) {
```

Comment:
The name of this function matters, there should be a comment on its definition saying that if you rename/renamespace it, you MUST update the static analysis too.

Reply:
I parameterized these strings.